modernpython/langPack.py

```diff
@@ -0,0 +1,156 @@
+import glob
+import logging
+import os
+import re
+import sys
+import textwrap
+from pathlib import Path
+
+import chevron
+
+logger = logging.getLogger(__name__)
+
+
+# This makes sure we have somewhere to write the classes, and
+# creates a couple of files the python implementation needs.
+# cgmes_profile_info details which uri belongs in each profile.
+# We don't use that here because we aren't creating the header
+# data for the separate profiles.
+def setup(version_path, cgmes_profile_info): # NOSONAR
+    if not os.path.exists(version_path):
+        os.makedirs(version_path)
+        _create_init(version_path)
+        _create_base(version_path)
+
+
+def location(version):
+    return "cimpy." + version + ".Base"
+
+
+base = {"base_class": "Base", "class_location": location}
+
+template_files = [{"filename": "cimpy_class_template.mustache", "ext": ".py"}]
+
+
+def get_class_location(class_name, class_map, version):
+    # Check if the current class has a parent class
+    if class_map[class_name].superClass():
+        if class_map[class_name].superClass() in class_map:
+            return "cimpy." + version + "." + class_map[class_name].superClass()
+        elif class_map[class_name].superClass() == "Base" or class_map[class_name].superClass() == None:
+            return location(version)
+    else:
+        return location(version)
+
+
+partials = {}
+
+
+# called by chevron, text contains the label {{dataType}}, which is evaluated by the renderer (see class template)
+def _set_default(text, render):
+    return _get_type_and_default(text, render)[1]
+
+
+def _set_type(text, render):
+    return _get_type_and_default(text, render)[0]
+
+
+def _get_type_and_default(text, renderer) -> tuple[str, str]:
+    result = renderer(text)
+    # the field {{dataType}} either contains the multiplicity of an attribute if it is a reference or otherwise the
+    # datatype of the attribute. If no datatype is set and there is also no multiplicity entry for an attribute, the
+    # default value is set to None. The multiplicity is set for all attributes, but the datatype is only set for basic
+    # data types. If the data type entry for an attribute is missing, the attribute contains a reference and therefore
+    # the default value is either None or [] depending on the multiplicity. See also write_python_files
+    # The default will be copied as-is, hence the possibility to have default or default_factory.
+    if result in ["M:1", "M:0..1", "M:1..1", ""]:
+        return ("Optional[str]", "default=None")
```

Comment:
If think I understand why you don't have recursion errors: you are not using other dataclassess as attribute type.

Albeit this avoids any recursion and make validation faster, it does not exploit the full power of pydantic, and makes difficult to navigate the data.